win32ss/gdi/ntgdi/freetype.c

```diff
@@ -220,7 +220,8 @@ FontLink_Chain_Free(
     if (!FontLink_Chain_IsPopulated(pChain)) // The chain is not populated yet
         return;
 
-    ExFreePoolWithTag(pChain->pszzFontLink, TAG_FONT);
+    if (pChain->pszzFontLink)
+        ExFreePoolWithTag(pChain->pszzFontLink, TAG_FONT);
```

Comment:
Shouldn't it just not add a font link into the chain if the `pszzFontLink` was not allocated? (cc Katayama for deeper investigation)

Reply:
[freetype.patch](https://github.com/user-attachments/files/16170107/freetype.patch)

@oleg-dubinskiy Can you try this patch? Thanks.

Reply:
@katahiromz just tested. Unfortunately your patch does not work yet.
No mentioned bugcheck anymore, but now I'm getting a last-chance exception instead, with the following stack trace:
```Entered debugger on last-chance exception (Exception Code: 0xc0000005) (Page Fault)
Memory at 0x00000008 could not be accessed

kdb:> bt
[?7h[cEip:
<win32k.sys:cc655 (win32ss/gdi/ntgdi/freetype.c:513 (FontLink_Chain_FindGlyph))>
Frames:
<win32k.sys:ce24a (win32ss/gdi/ntgdi/freetype.c:5026 (TextIntGetTextExtentPoint))>
<win32k.sys:f061f (win32ss/gdi/ntgdi/text.c:417 (NtGdiGetTextExtentExW))>
<ntoskrnl.exe:3fe5 (:0 (KiSystemCallTrampoline))>
<ntoskrnl.exe:168bb1 (ntoskrnl/ke/i386/traphdlr.c:1840 (KiSystemServiceHandler))>
<ntoskrnl.exe:3e2f (:0 (KiFastCallEntry))>
<ntdll.dll:f8ad>
<lpk.dll:34fa>
<gdi32.dll:18dff>
<user32.dll:6a1fb>
<user32.dll:4e812>
<DiscordMessenger.exe:9096e>
<user32.dll:6b871>
<user32.dll:5fb67>
<user32.dll:5ff81>
<user32.dll:60b7a>
<COMCTL32.DLL:4ddb6>
<COMCTL32.DLL:4f3eb>
<COMCTL32.DLL:57849>
<user32.dll:6b871>
<user32.dll:5fb67>
<user32.dll:60111>--- Press q to abort, any other key to continue ---

<DiscordMessenger.exe:8fc49>
<user32.dll:5fb67>
<user32.dll:619ae>
<ntdll.dll:f800>
<DiscordMessenger.exe:8a65e>
Couldn't access memory at 0x000002C0!

kdb:>
```

See full 
[debug.log](https://github.com/user-attachments/files/16172787/debug.log)
 too.
It occurs just after login in the app, when it tries to redraw the fonts first time, not even when switching between the channels:
![dm-hang](https://github.com/reactos/reactos/assets/26385117/c9406b6e-2535-4008-b3c8-98900b609f99)
If to try continue an execution, the following another bugcheck occurs instead, with the following stack trace:

```
*** Fatal System Error: 0x0000001e
                       (0xC0000005,0xF654D655,0xEDF495C8,0x00000000)


Entered debugger on embedded INT3 at 0x0008:0x805AB3AC.

kdb:> bt
Eip:
<ntoskrnl.exe:1ab3ac (sdk/lib/rtl/i386/debug_asm.S:56 (RtlpBreakWithStatusInstruction))>
Frames:
<ntoskrnl.exe:97dcd (ntoskrnl/ke/bug.c:1067 (KeBugCheckWithTf))>
<ntoskrnl.exe:98337 (ntoskrnl/ke/bug.c:1416 (KeBugCheckEx))>
<ntoskrnl.exe:162bd8 (ntoskrnl/ke/i386/exp.c:888 (KiDispatchException))>
<ntoskrnl.exe:163006 (ntoskrnl/ke/i386/exp.c:1081 (KiDispatchExceptionFromTrapFrame))>
<ntoskrnl.exe:167edd (ntoskrnl/include/internal/i386/ke.h:836 (KiTrap0EHandler))>
<ntoskrnl.exe:36ae (:0 (KiTrap0E))>
<win32k.sys:cc650 (win32ss/gdi/ntgdi/freetype.c:4406 (FontLink_Chain_FindGlyph))>
<win32k.sys:ce24a (win32ss/gdi/ntgdi/freetype.c:5026 (TextIntGetTextExtentPoint))>
<win32k.sys:f061f (win32ss/gdi/ntgdi/text.c:417 (NtGdiGetTextExtentExW))>
<ntoskrnl.exe:3fe5 (:0 (KiSystemCallTrampoline))>
<ntoskrnl.exe:168bb1 (ntoskrnl/ke/i386/traphdlr.c:1840 (KiSystemServiceHandler))>
<ntoskrnl.exe:3e2f (:0 (KiFastCallEntry))>
<ntdll.dll:f8ad>
<lpk.dll:34fa>
<gdi32.dll:18dff>
<user32.dll:6a1fb>
<user32.dll:4e812>
<DiscordMessenger.exe:9096e>
<user32.dll:6b871>
<user32.dll:5fb67>
<user32.dll:5ff81>--- Press q to abort, any other key to continue ---

<user32.dll:60b7a>
<COMCTL32.DLL:4ddb6>
<COMCTL32.DLL:4f3eb>
<COMCTL32.DLL:57849>
<user32.dll:6b871>
<user32.dll:5fb67>
<user32.dll:60111>
<DiscordMessenger.exe:8fc49>
<user32.dll:5fb67>
<user32.dll:619ae>
<ntdll.dll:f800>
<DiscordMessenger.exe:8a65e>
Couldn't access memory at 0x000002C0!

kdb:>
```

Here's a new [debug.log](https://github.com/user-attachments/files/16172881/debug.log) with the bugcheck included too.

Reply:
> Shouldn't it just not add a font link into the chain if the `pszzFontLink` was not allocated? (cc Katayama for deeper investigation)

@binarymaster it will not change anything, because in that case, `pChain->pszzFontLink` will be initialized to `NULL` (we can't just keep it uninitialized), so the check will still be needed.

Reply:
Here is my personal analysis of the code.

I think the current PR (by Oleg) is part of the correct fix. Indeed, as the current font-link code stands, the reason why a `PFONTLINK_CHAIN` object can already exist by itself without its `pszzFontLink` allocated, is because it is delay-populated.

1. Given a `FONTLINK_CHAIN Chain` (containing a list of font-links objects, corresponding to the contents of the textual `pszzFontLink`), such chain object [can be created](https://git.reactos.org/?p=reactos.git;a=blob;f=win32ss/gdi/ntgdi/freetype.c;hb=465b9cef25b6d57451f6b2b80c29275d75b27734#l5006) as a _**local object**_, but stays "empty", until a [font lookup is made](https://git.reactos.org/?p=reactos.git;a=blob;f=win32ss/gdi/ntgdi/freetype.c;hb=465b9cef25b6d57451f6b2b80c29275d75b27734#l5025), for example by `FontLink_Chain_FindGlyph(&Chain, ...)`.
(Note that many successive calls to `FontLink_Chain_FindGlyph(&Chain, ...)` can be made using the same Chain object.)
If `FontLink_Chain_FindGlyph()` finds the chain to be [not yet populated](https://git.reactos.org/?p=reactos.git;a=blob;f=win32ss/gdi/ntgdi/freetype.c;hb=465b9cef25b6d57451f6b2b80c29275d75b27734#l4395), it calls `FontLink_Chain_Populate()` on the Chain.

2. The `FontLink_Chain_Populate()` function initializes a font face name for the chain (which is used later as the [criterium to know whether a font chain is populated by `FontLink_Chain_IsPopulated()`](https://git.reactos.org/?p=reactos.git;a=blob;f=win32ss/gdi/ntgdi/freetype.c;hb=465b9cef25b6d57451f6b2b80c29275d75b27734#l207), and is also used here by `FontLink_Chain_Free()`...), then does some bookkeeping and [calls `FontLink_Chain_LoadReg()`](https://git.reactos.org/?p=reactos.git;a=blob;f=win32ss/gdi/ntgdi/freetype.c;hb=465b9cef25b6d57451f6b2b80c29275d75b27734#l1523).

3. It is that latter `FontLink_Chain_LoadReg()` function that loads the font-link data from the registry and inititializes, if possible, the `pChain->pszzFontLink` member [to an allocated buffer](https://git.reactos.org/?p=reactos.git;a=blob;f=win32ss/gdi/ntgdi/freetype.c;hb=465b9cef25b6d57451f6b2b80c29275d75b27734#l1387).
If the `FontLink_Chain_LoadReg()` function fails, no `pszzFontLink` is allocated, i.e. `pChain->pszzFontLink == NULL`; _**however**_ one can note that such chain can still be marked as "populated" because a font face name was correctly set in it by `FontLink_Chain_Populate()` (see point 2).
Note that [if `FontLink_Chain_LoadReg()` fails, `FontLink_Chain_Populate()` fails as well](https://git.reactos.org/?p=reactos.git;a=blob;f=win32ss/gdi/ntgdi/freetype.c;hb=465b9cef25b6d57451f6b2b80c29275d75b27734#l1523).
However, the return value of `FontLink_Chain_Populate()` is [ignored in the call made to it by `FontLink_Chain_FindGlyph()`](https://git.reactos.org/?p=reactos.git;a=blob;f=win32ss/gdi/ntgdi/freetype.c;hb=465b9cef25b6d57451f6b2b80c29275d75b27734#l4396).

4. Back to the caller of `FontLink_Chain_FindGlyph()` that created the `Chain` object as well: once the lookup has been done and the `Chain` object isn't needed anymore, it is either discarded or added into the font-link cache by a call to [`FontLink_Chain_Finish(&Chain)`](https://git.reactos.org/?p=reactos.git;a=blob;f=win32ss/gdi/ntgdi/freetype.c;hb=465b9cef25b6d57451f6b2b80c29275d75b27734#l5070).

5. If the chain isn't populated, nothing is done, otherwise it is [actually copied to a persistent structure and linked into the cache](https://git.reactos.org/?p=reactos.git;a=blob;f=win32ss/gdi/ntgdi/freetype.c;hb=465b9cef25b6d57451f6b2b80c29275d75b27734#l272).
### This is where a "populated" chain object that _**can have an empty non-allocated `pszzFontLink` member**_ (due to a failure in `FontLink_Chain_LoadReg()` and `FontLink_Chain_Populate()`) _**gets anyway linked into the cache**_.

6. Much later, in case a cached chain needs to be freed via [`FontLink_Chain_Free()`](https://git.reactos.org/?p=reactos.git;a=blob;f=win32ss/gdi/ntgdi/freetype.c;hb=465b9cef25b6d57451f6b2b80c29275d75b27734#l213), such a linked chain appears to _**always report to be populated**_ (i.e. the `if (!FontLink_Chain_IsPopulated(pChain)) // The chain is not populated yet` test will always fail, because such chain, added into the cached list by `FontLink_Chain_Finish()`, as we saw above, is always added only if it reports as populated).
_**BUT**_, its `pszzFontLink` member may still be NULL, if such a Chain was linked into the cache.
Thus, _**at present, a check for `pszzFontLink` being non-NULL needs to be added before freeing it**_, as done by this PR.

Note for @katahiromz : it seems that the the `if (!FontLink_Chain_IsPopulated(pChain)) // The chain is not populated yet` test in `FontLink_Chain_Free()` is never going to be hit in principle, and could be turned into an `ASSERT(FontLink_Chain_IsPopulated(pChain));` instead.

Reply:
OK, so the rest of the correct fix can be committed later, and my part is OK to be merged? Or everything necessarily should be made in one PR? :thinking: